File: web/playwright-java/src/test/java/utils/ResilientLocator.java
package utils;

import com.microsoft.playwright.*;

/**
 * Self-healing locator — tries data-test-id → role → text → CSS.
 */
public class ResilientLocator {
    private final Page page;
    private final double timeout;

    public ResilientLocator(Page page) {
        this(page, 5000);
    }

    public ResilientLocator(Page page, double timeout) {
        this.page = page;
        this.timeout = timeout;
    }

    public Locator find(String testId, String role, String roleName, String text, String css) {
        double per = timeout / 4;
        if (testId != null) {
            Locator loc = page.locator("[data-test-id=\"" + testId + "\"]").first();
            if (loc.isVisible(new Locator.IsVisibleOptions().setTimeout(per))) return loc;
        }
        if (role != null) {
            Locator loc = page.getByRole(AriaRole.valueOf(role.toUpperCase()),
                    new Page.GetByRoleOptions().setName(roleName)).first();
            if (loc.isVisible(new Locator.IsVisibleOptions().setTimeout(per))) return loc;
        }
        if (text != null) {
            Locator loc = page.getByText(text).first();
            if (loc.isVisible(new Locator.IsVisibleOptions().setTimeout(per))) return loc;
        }
        if (css != null) {
            return page.locator(css).first();
        }
        return page.locator("[data-test-id=\"" + testId + "\"]");
    }
}

File: web/playwright-python/utils/resilient_locator.py
from playwright.sync_api import Page, Locator


class ResilientLocator:
    """Self-healing locator — tries data-test-id → role → text → CSS."""

    def __init__(self, page: Page, timeout: float = 5000):
        self.page = page
        self.timeout = timeout

    def find(self, *, test_id: str = None, role: str = None, role_name: str = None,
             text: str = None, css: str = None) -> Locator:
        per = self.timeout / 4

        if test_id:
            loc = self.page.locator(f'[data-test-id="{test_id}"]').first
            if loc.is_visible(timeout=per):
                return loc

        if role:
            loc = self.page.get_by_role(role, name=role_name).first
            try:
                if loc.is_visible(timeout=per):
                    return loc
            except Exception:
                pass

        if text:
            loc = self.page.get_by_text(text).first
            try:
                if loc.is_visible(timeout=per):
                    return loc
            except Exception:
                pass

        if css:
            return self.page.locator(css).first

        return self.page.locator(f'[data-test-id="{test_id}"]')
